src/Rules/Functions/ImplodeFunctionRule.php

```diff
@@ -0,0 +1,64 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Functions;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\UnionType;
+
+/**
+ * @implements \PHPStan\Rules\Rule<\PhpParser\Node\Expr\FuncCall>
+ */
+class ImplodeFunctionRule implements \PHPStan\Rules\Rule
+{
+
+	private \PHPStan\Reflection\ReflectionProvider $reflectionProvider;
+
+	public function __construct(
+		ReflectionProvider $reflectionProvider
+	)
+	{
+		$this->reflectionProvider = $reflectionProvider;
+	}
+
+	public function getNodeType(): string
+	{
+		return FuncCall::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (!($node->name instanceof \PhpParser\Node\Name)) {
+			return [];
+		}
+
+		$functionName = $this->reflectionProvider->resolveFunctionName($node->name, $scope);
+		if (!in_array($functionName, ['implode', 'join'], true)) {
+			return [];
+		}
+
+		$args = $node->getArgs();
+		if (count($args) === 1) {
+			$arrayType = $scope->getType($args[0]->value);
+		} elseif (count($args) === 2) {
+			$arrayType = $scope->getType($args[1]->value);
+		} else {
+			return [];
+		}
+
+		if ($arrayType->getIterableValueType()->toString() instanceof ErrorType) {
```

Comment:
We should also employ RuleLevelHelper::findTypeToCheck so that partially-wrong unions aren't reported until level 7.

Ther are many usages of that method, look through the codebase.

Reply:
adjusted. seems to work. does this require some kind of unit-test?

Reply:
Check out LevelsIntegrationTest, add something in acceptTypes.php, and re-run this test case. The JSON snapshots will be updated.

Reply:
I guess it works now.